File: 0328-odd-even-linked-list/0328-odd-even-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (head == nullptr) return head;
        
        ListNode* curNode = head;
        ListNode* tmpNode, *prevNode;
        ListNode* evenStart=head-> next;
        int i = 0;
        // connecting node to the next next node
        while(curNode->next != nullptr){ // loop over until curNode is not the last elem of linked list
            tmpNode=curNode -> next;
            curNode->next = tmpNode-> next; // connect current node to a node 2 steps ahead
            prevNode = curNode; // save current curNode for later
            curNode = tmpNode; // move curNode 1 step head
            i++;
        }
        // connecting end of odd group to start of even group
        if (i%2==0) curNode -> next = evenStart;  // if the linked list length is odd
        else prevNode->next = evenStart; // if the linked list length is even
        return head;
        
    }
};

File: linked-list/0328-odd-even-linked-list/0328-odd-even-linked-list.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return head
        else:
            curNode = head
            tmpNode = None
            prevNode = None
            i = 0
            evenStart = None
            while curNode.next != None:
                if i + 1 == 1:
                    evenStart = curNode.next
                tmpNode = curNode.next
                curNode.next = tmpNode.next
                prevNode = curNode
                curNode = tmpNode
                i += 1
            if i % 2 == 0:
                curNode.next = evenStart
            else:
                prevNode.next = evenStart
            return head